### DISCUSSION TITLE
Public entity properties

### DISCUSSION THREAD

**Discussion by OliverSkroblin** [shopware/platform#3166] [RFC / Proposal]
**effort**: middle

### Description
Currently the type check of entity classes is done via the getter and setter. However, with php 7.4 typed properties were introduced, making the getter and setter obsolete in many cases.

Therefore we want to remove these unnecessary methods from the entities. Benchmarks have already shown that we can optimize the hydration performance. 

Furthermore, this also reduces the LOC, which must be maintained and tested by us.

Getters and setters of properties that are to be protected for the app scripts remain in place in order to be able to continue to restrict access.

### Benefits
* Less boiler plate code
* Performance improvement in the area of entity hydration

### Break strategy
* We would already set all affected properties to public. 
* We will deprecate the affected getter and setter functions with `@deprectated tag:v6.5.0`
* With the major version we will then remove all functions.
* Since this would cause an enormous effort for third party developers, we implement the `__call` method in the entity class to continue to support the getter and setter calls.

**Comment by aragon999** [comment]
I am very much in favor, of this proposal, just to avoid that the `__call` method will be permanent, it probably should also be deprecated for the next major version, or maybe 6.6.0 :)

**Comment by windaishi** [comment]
I'd appreciate this for properties that are not an association or foreign key. 

**I do NOT recommend this for properties that are associations or foreign keys.**

In our code we don't have just simple getters and setters for associations and foreign keys but usually have this kind of convenience code:

```php
class ShipmentEntity extends Entity
{
    use EntityIdTrait;

    // The sales channel is optional
    protected ?string $salesChannelId = null;
    protected ?SalesChannelEntity $salesChannel = null;

    public function getSalesChannel(): ?SalesChannelEntity
    {
        if ($this->salesChannel === null && $this->salesChannelId !== null) {
            throw new AssociationNotLoadedException('salesChannel', $this);
        }

        return $this->salesChannel;
    }

    public function setSalesChannel(?SalesChannelEntity $salesChannel): void
    {
        $this->salesChannel = $salesChannel;
        if ($salesChannel !== null) {
            $this->salesChannelId = $salesChannel->getId();
        }
    }

    public function getSalesChannelId(): ?string
    {
        return $this->salesChannelId;
    }

    public function setSalesChannelId(?string $salesChannelId): void
    {
        $this->salesChannelId = $salesChannelId;
        if ($salesChannelId === null
            || ($this->salesChannel !== null && $this->salesChannel->getId() !== $salesChannelId)
        ) {
            $this->salesChannel = null;
        }
    }
}
```
(Example code for a `ManyToOneAssociation`. We have similar code for other type of associations.)

This code helps us to not accidentally assume associations to be null just because we forgot to load it. Instead we get an exception when we try to access it but it has not been loaded.

If the properties for associations would be public we could not enforce that this method would be called and therefore risk nasty bugs.

## So instead I suggest:
* Make only non-association properties public
* For properties that belong to an associations just code like the above in all entities.

We can provide code for other kind of associations.

-----
**Edit:**

Maybe the "initialized property" feature of PHP 7.4 (that came with typed properties) could be used instead of the suggested getter/setter code instead.

An example would be:
```php
<?php
class Entity {}
class ShipmentEntity extends Entity
{
    public ?string $salesChannelId = null;
    public SalesChannelEntity $salesChannel;
}

$shipment = new ShipmentEntity();
// Sales channel was not loaded therefore the property is not initialized:
var_dump($shipment->salesChannel);
// Fatal error: Uncaught Error: Typed property ShipmentEntity::$salesChannel must not be accessed before initialization in /in/CPlqC:11
```

https://3v4l.org/brmSM

You can also "reset" a property with `unset()`: https://3v4l.org/maRST

We should put research in this topic...

**Comment by AndreasA** [comment]
Normally I would keep the option to use getters/setters for developers as setters could also verify data (e.g. that is transferred into the database like only support values `> 0 `.

However, with Shopware 6 and DAL the data that is to be written is validated differently and therefore it is fine. For any validation the field serializer's or entity loaded events could still be used.
For adding additional data, the corresponding entity loaded event exists and should be used anyway instead of doing so through setters.
And if associations should not be nullable, the property uninialized exception can be used, so I think for Shopware 6 all use cases can be handled.

----

However, what might be even better would be with PHP 8.1 to use readonly properties in combination with named parameters on the constructor. That way the properties are initialized once, accessible for read from outside but cannot be modified (or unset) from the outside.
Or at least use the constructor to initialize the properties.

**EDIT:** Though - now that I thought about it in greater detail - I guess there are actually a lot of scenarios where they should be modifyable, e.g. using entity loaded events etc. could be used to change a product's name which could depend on a sales channel config (although one could also create a new object then. However, some properties like `id` most likely are only readable.

example:

```php
<?php declare(strict_types=1);

class MyEntity {
    public function __construct(
        public readonly string $id,
        public readonly string $test,
    ) {}
}

$entity = new MyEntity(test: 'foo', id: 'bar');

var_dump($entity->id);
```
Result: `string(3) "bar"`

or with dynamic argument list:

```php
<?php declare(strict_types=1);

class MyEntity {
    public function __construct(
        public readonly string $id,
        public readonly string $test,
    ) {}
}

$args = ['test' => 'foo', 'id' => 'bar'];

$entity = new MyEntity(...$args);

var_dump($entity->id);
````
Result `string(3) "bar"`
